[RFC] ath79: Fix/update DTS for TP-Link Archer A7/C7 v5

adschm opened the pull request:
During the support for the Archer C7 v4 some possible improvements for the C7 v5 have been indicated [1].

This PR is supposed to collect some (hopefully all) of them.

> It is worth mentioning that Archer C7 v5 DTS most probably should be fixed according to this pull request, because its DTS contains the following issues:
> 
> * There is only one LED alias `led-status`, so I doubt that anything blinks during boot indicating correct boot stage
> * qca,ar8327-initvals:
>   
>   * it initializes LEDS, but if v5 is same as v4, there are no ethernet LEDS on the backpanel, so it is meaningless to initialize them
>   * it does not initialize PORT6, so it will look like connected - ar71xx code initializes also PORT6 and it appears as disconnected (I have tested this on v4)
>   * if v5 also uses QCA8337N, the PORT0 PAD CTRL init value on ath79 is different to one used during init on ar71xx, where the highest bit is set (bit MAC06_EXCHANGE_EN). I cannot say what this change means, though, but my intention was to have v4 init values aligned (ath79 uses the same values as ar71xx).

I would be glad about some help with the qca,ar8327-initvals stuff.

Will be able to test on a C7 v5 soon. Feedback from A7 v5 would be glad.

Done:
* Merge LEDS
* led-status
* Update reset button GPIO
* Fix WAN-MAC for ath79
* Fix WAN-MAC for ar71xx (requires to add info partition first)

ToDo:
* check/fix qca,ar8327-initvals

[1] https://github.com/openwrt/openwrt/pull/1726#issuecomment-455096720

adschm committed:
ath79: Consolidate LEDs in Archer A7/C7 v5 DTSI

Definition is split here without obvious reason. Just merge it
(and align order to that from C7 v4).

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ath79: Utilize new LED modes from diag.sh for Archer A7/C7 v5

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
@howl @oldium @TOCK-Chiu @k1ll

adrianschmutzler commented:
In ar71xx, the reset button GPIO for the C7v5 is "5" (https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-archer-c7-v5.c#L45).

In the ath79 DTSI, it is "11". (https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_archer-x7-v5.dtsi#L47)

Is this wrong then? Is it also wrong for A7v5?

adrianschmutzler commented:
> it initializes LEDS, but if v5 is same as v4, there are no ethernet LEDS on the backpanel, so it is meaningless to initialize them

I do not see any LEDs on the back of my C7v5. Does that mean the four lines can be dropped?

981213 commented:
ar8327-initvals sets switch registers. You can find leaked switch datasheet here:
https://github.com/Deoptim/atheros

adrianschmutzler commented:
@981213 Thanks for the link. I will have a look.

Maybe you can help me with another topic, since you seem to be into this stuff based on your other PRs:
When looking at the DTS files of C7v4/v5, I found that that eth1 is disabled, and all ports in the switch are linked to eth0 (02_network config file). However, to me it looks more reasonable to use both eth0 and eth1 for LAN and WAN, as it is done e.g. for the C7v2.
With this change, WAN would get its dedicated eth, MAC address could be set directly in DTS etc.

Do you have an idea whether this is possible and may you provide me some insight into that or even hint about moving the C7v5 into this direction?
(Same could then be leveraged for the C7v4...)

C7v5:
https://github.com/oldium/openwrt-openwrt/blob/master/target/linux/ath79/dts/qca956x.dtsi#L301
https://github.com/oldium/openwrt-openwrt/blob/master/target/linux/ath79/base-files/etc/board.d/02_network#L163

C7v2:
https://github.com/oldium/openwrt-openwrt/blob/master/target/linux/ath79/dts/qca9558_tplink_archer-c7.dtsi#L163
https://github.com/oldium/openwrt-openwrt/blob/master/target/linux/ath79/base-files/etc/board.d/02_network#L143

adschm committed:
ar71xx: Correct MAC address for WAN interface of Archer C7 v5

This device shares the network config with v4, thus the WAN MAC
also needs to be fixed the same way. However, the partition
where the MAC address resides has been changed.

Based on: https://github.com/openwrt/openwrt/pull/1726

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ath79: Correct MAC address for WAN interface of Archer A7/C7 v5

These devices share the network config with C7v4, thus the WAN MAC
also needs to be fixed the same way. However, the partition
where the MAC address resides has been changed.

Based on: https://github.com/openwrt/openwrt/pull/1726

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

981213 commented:
QCA955X has 2 exposed GMACs, one for RGMII and another for SGMII. In Archer C7 v2 case, both of them are connected to AR8327.
QCA9563, however, only have one SGMII on GMAC0. The other GMAC doesn't exist. On QCA9561 that second GMAC connects to built-in switch.

k1ll commented:
> In ar71xx, the reset button GPIO for the C7v5 is "5" (https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-archer-c7-v5.c#L45).
> 
> In the ath79 DTSI, it is "11". (https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_archer-x7-v5.dtsi#L47)
> 
> Is this wrong then? Is it also wrong for A7v5?

Just tested again to make sure and GPIO "11" on my A7v5 resets the device. GPIO "5" seems to do nothing. I don't have a C7v5 so i can't check what is correct on that device.

adrianschmutzler commented:
@k1ll Thanks for testing. I will soon be able to test ar71xx and ath79 on the C7v5 ... Let's see.

adrianschmutzler commented:
> Just tested again to make sure and GPIO "11" on my A7v5 resets the device. GPIO "5" seems to do nothing. I don't have a C7v5 so i can't check what is correct on that device.

Test results from C7v5: Reset on ar71xx works, reset on ath79 does not do anything.

So that looks like the GPIO is actually different between A7 and C7. All other values matches those from ar71xx, however. I will do a patch for this (separating A7 and C7), test it and then push it here.

ynezz commented:
@adrianschmutzler I'm wondering if there is any reason for this cross-posting (GitHub and mailing list) ? It would be nice if you could stick to one platform for all this fixes/improvements and take care of your patches in the patchwork as well (I've removed those two for MAC address fixes from patchwork already). Thanks.

adschm committed:
ath79: Fix GPIO reset button on TP-Link Archer C7v5

The GPIO for the reset button for the Archer C7v5 changed from
ar71xx to ath79. An investigation based on tests revealed
that the A7v5 responds on "11", while the C7v5 responds on
"5" as set for ar71xx.

Thus, we just define this in the DTS files instead of in the
common DTSI.

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
I added the GPIO reset button fix and the corrected patch for the WAN-MAC on ath79 (as ynezz righteously noted).

Successfully tested on C7v5:
- General LED functionality
- Boot LED status updates
- Reset works
- WAN-MAC calculation works

I will also push the ar71xx WAN patch here later, when I've corrected and tested it.
Same for the ar8327-initvals, after having understood and evaluated this one.

Would be glad if @k1ll could give it some testing on the A7v5, even now or after the rest has been pushed. Do you now whether the stock-FW WAN address for the A7v5 actually is LANMAC+1, as for the C7v5?

adschm committed:
ar71xx: Add "info" partition for TP-Link Archer C7 v5

This adds the "info" MTD partition, as it is specified in the
ath79 DTS:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_archer-c7-v5.dts#L35

This is required to set the WAN MAC address, as it is build based
on the LAN MAC address, which in turn has to be read from the
"info" partition:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_archer-c7-v5.dts#L35

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -319,6 +319,11 @@ ath79_setup_macs()
+        tplink,archer-c7-v5)
+                base_mac=$(mtd_get_mac_binary info 8)
+                wan_mac=$(macaddr_add "$base_mac" 1)
+                ;;
```
Leading spaces here, will fix this when I push the ar71xx stuff.

k1ll commented:
> I added the GPIO reset button fix and the corrected patch for the WAN-MAC on ath79 (as ynezz righteously noted).
> 
> Successfully tested on C7v5:
> 
>     * General LED functionality
> 
>     * Boot LED status updates
> 
>     * Reset works
> 
>     * WAN-MAC calculation works
> 
> 
> I will also push the ar71xx WAN patch here later, when I've corrected and tested it.
> Same for the ar8327-initvals, after having understood and evaluated this one.
> 
> Would be glad if @k1ll could give it some testing on the A7v5, either now or after the rest has been pushed. Do you now whether the stock-FW WAN address for the A7v5 actually is LANMAC+1, as for the C7v5?

Yes stock-FW WAN address is LANMAC+1 on A7v5 as well. Tested the patches on my A7v5 and everything works as expected.

adrianschmutzler commented:
@ynezz Dealing with the qca,ar8327-initvals will challenge me and I do not know whether I will have time for that soon. Thus, feel free to merge what's already here (which is all tested).

ynezz commented on target/linux/ar71xx/image/generic-tp-link.mk:
```diff
@@ -154,7 +154,7 @@ define Device/archer-c7-v5
   TPLINK_BOARD_ID := ARCHER-C7-V5
   IMAGE_SIZE := 15360k
-  MTDPARTS := spi0.0:128k(factory-uboot)ro,128k(u-boot)ro,64k@0x50000(art)ro,15360k@0xc0000(firmware)
+  MTDPARTS := spi0.0:128k(factory-uboot)ro,128k(u-boot)ro,64k@0x50000(art)ro,128k@0x60000(info)ro,15360k@0xc0000(firmware)
```
I'm wondering why is this needed. Please can you point me to some place where it's actually being used?

adrianschmutzler replied:
Unfortunately I placde the same link twice in the commit description. The second one should be:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_archer-x7-v5.dtsi#L184

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

adrianschmutzler commented:
@ynezz Thanks for pulling this one.
If you want, you may still update the second link in the commit message of my patch "ar71xx: Add "info" partition for TP-Link Archer C7 v5 ":

> Unfortunately I placed the same link twice in the commit description. The second one should be:
> https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_archer-x7-v5.dtsi#L184

However, if you don't, it won't be a terrible issue.

ynezz commented:
> @ynezz Dealing with the qca,ar8327-initvals will challenge me and I do not know whether I will have time for that soon. Thus, feel free to merge what's already here (which is all tested).

I do not know which ar8327 regs you were referring to, but I would like to get the LEDs removed in https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=810397ef120d5013015e0d4aba72da4c940449fe as the LEDs are not connected to the switch, but rather to the SoC.

adrianschmutzler commented:
@ynezz I was referring to the discussion from C7v4 (https://github.com/openwrt/openwrt/pull/1726#issuecomment-455096720):

> qca,ar8327-initvals:
> 
> it initializes LEDS, but if v5 is same as v4, there are no ethernet LEDS on the backpanel, so it is meaningless to initialize them
> it does not initialize PORT6, so it will look like connected - ar71xx code initializes also PORT6 and it appears as disconnected (I have tested this on v4)
> if v5 also uses QCA8337N, the PORT0 PAD CTRL init value on ath79 is different to one used during init on ar71xx, where the highest bit is set (bit MAC06_EXCHANGE_EN). I cannot say what this change means, though, but my intention was to have v4 init values aligned (ath79 uses the same values as ar71xx).

So their opinion also was that the LED initialization should be removed. If you come to the same conclusion, it probably right to do so.

Didn't have time to look into the PORT0/PORT6 init values so far.

ynezz commented:
> Didn't have time to look into the PORT0/PORT6 init values so far.

https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=aa655a146547016f2d264ac1485a4e9d9bbf49cf

adrianschmutzler commented:
@ynezz : I have run-tested both your qca,ar8327-initvals patches (with the MIB counter stuff).

I do not see any noteable difference.

ynezz commented:
Adrian Schmutzler <notifications@github.com> [2019-05-05 11:45:22]:

> @ynezz : I have run-tested both your qca,ar8327-initvals patches (with the MIB counter stuff).
> 
> I do not see any noteable difference.

Thanks, so to which patches exactly could I add your `Tested-by` ?

adrianschmutzler commented:
https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=4e9dd7d420757e96935f4e8e5a796d424e2346ea
https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=d442f51f78702689e91fff144e2a012487a8f6fe
I only tested those two based on current master (WITHOUT the MIB counter stuff)
I did NOT evaluate whether the set initvals are reasonable based on the "manual".

Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
Just had some time for a look into the docs:
I'm still puzzled by the PORT6_STATUS Register
The value for the C7v5 from ar71xx (in your patch) is 0x00001080.
This will set FLOW_LINK_EN_6 and TX_HALF_FLOW_EN_6.
For the C7v4, they set 0x00000200 (as in ar71xx) and reported that port 6 shows as disconnected (as they expected, [1]). This is equivalent to LINK_EN_6 enabled.
Which is correct here (assuming that both are similar and only one is correct)?

[1] https://github.com/openwrt/openwrt/pull/1726#issuecomment-455096720

oldium commented:
Just a background on how I found values for DTS for C7v4 - I added logging to kernel in the ar71xx path to know which values are actually used during initialization and simply used the same values in DTS for ath79. Then I checked the differences from C7v5 and tried to find out some (little) information about it.

ynezz commented:
> Just a background on how I found values for DTS for C7v4 - I added logging to kernel in the ar71xx path to know which values are actually used during initialization and simply used the same values in DTS for ath79.

I did the same, from my notes:

```
   3         2         1
  10987654321098765432109876543210
0b10000000000010000000000010000000 PORT0_PAD_CTRL
0b10000001000000000000000010000000 POWER_ON_STRAP (PWS_REG)
0b00000000000000000000000001111110 PORT0_STATUS
0b00000000000000000001000010000000 PORT6_STATUS

PORT0_PAD_CTRL |= MAC0_SGMII_EN(7) | SGMII_CLK125M_RX_SEL(19)
POWER_ON_STRAP |= SERDES_AEN(7) | LED_OPEN_EN_CSR (24) | POWER_ON_SEL (31)

PORT0_STATUS   |= SPEED_0(10) | TXMAC_EN0(2) | RXMAC_EN0(3) | TX_FLOW_EN0(4)
                    1000M
               |= RX_FLOW_EN0(5) | DUPLEX_MODE0(6)
                                   Full-duplex mode

PORT6_STATUS   |= TX_HALF_FLOW_EN(7) | FLOW_LINK_EN(12)
```

adrianschmutzler commented:
Hmm, strange. I just ran three different images on my C7v5:

ar71xx, openwrt-18.06:
root@FF-BTL-C7V5:~# swconfig dev switch0 port 6 show
Port 6:
        mib: No MIB data
        enable_eee: ???
        igmp_snooping: 0
        vlan_prio: 0
        pvid: 0
        link: port:6 link:down

ath79, 0x00001080:
root@FF-BTL-C7V5:~# swconfig dev switch0 port 6 show
Port 6:
        mib: No MIB data
        enable_eee: ???
        igmp_snooping: 0
        vlan_prio: 0
        pvid: 0
        link: port:6 link:up speed:10baseT half-duplex

ath79, 0x00000200:
root@FF-BTL-C7V5:~# swconfig dev switch0 port 6 show
Port 6:
        mib: No MIB data
        enable_eee: ???
        igmp_snooping: 0
        vlan_prio: 0
        pvid: 0
        link: port:6 link:down

If I find time, I will also try ar71xx from master.

adrianschmutzler commented:
ar71xx, master (1 or 2 months old):
root@FF-BTL-C7V5:~# swconfig dev switch0 port 6 show
Port 6:
        mib: No MIB data
        enable_eee: ???
        igmp_snooping: 0
        vlan_prio: 0
        pvid: 0
        link: port:6 link:down

adrianschmutzler commented:
From this, I would take 0x00000200 to set port down as for C7v4. But I do not have the slightest idea why you didn't get this in the first place.

ynezz commented:
> From this, I would take 0x00000200 to set port down as for C7v4.

Yep, thanks for spotting this https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=091ebf4c7b7b73b1fa1aa375fe767affe77b243c

> But I do not have the slightest idea why you didn't get this in the first place.

Me neither, I've just run it again and got following output, quite strange, but it might be possible, that I've had simply some other patch in the play, it happens :(

```
[    0.753382] *** AR8327_REG_PAD0_MODE 0x80080080/0x80080080
[    0.760082] *** AR8327_REG_PAD5_MODE 0x0/0x0
[    0.765506] *** AR8327_REG_PAD6_MODE 0x0/0x0
[    0.828625] *** AR8327_REG_PORT_STATUS(0) 0x7e/0x7e
[    0.955741] *** AR8327_REG_PORT_STATUS(1) 0x1200/0x1200
[    1.075741] *** AR8327_REG_PORT_STATUS(2) 0x1200/0x1200
[    1.195740] *** AR8327_REG_PORT_STATUS(3) 0x1200/0x1200
[    1.315742] *** AR8327_REG_PORT_STATUS(4) 0x1200/0x1200
[    1.435741] *** AR8327_REG_PORT_STATUS(5) 0x1200/0x1200
[    1.444184] *** AR8327_REG_PORT_STATUS(6) 0x200/0x200
```

adrianschmutzler commented:
Since this is the result of a longer story, I would be a litte more chatty in the commit message and add:

"This will harmonize settings with the similar Archer C7 v4. Port 6 will appear as disconnected."

Despite that, I'm fine with the patch. :-)
Thanks for pulling this through ...
